File: src/main/java/com/pestphp/pest/PestFrameworkType.java
package com.pestphp.pest;

import com.jetbrains.php.testFramework.PhpTestDescriptor;
import com.jetbrains.php.testFramework.PhpTestFrameworkConfigurationFactory;
import com.jetbrains.php.testFramework.PhpTestFrameworkFormDecorator;
import com.jetbrains.php.testFramework.PhpTestFrameworkFormDecorator.PhpDownloadableTestFormDecorator;
import com.jetbrains.php.testFramework.PhpTestFrameworkType;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.List;

public class PestFrameworkType extends PhpTestFrameworkType {
    @NonNls
    public static final String ID = "Pest";

    @NotNull
    public static PhpTestFrameworkType getInstance() {
        return PhpTestFrameworkType.getTestFrameworkType(ID);
    }

    @Override
    public @NotNull @Nls String getDisplayName() {
        return PestBundle.message("FRAMEWORK_NAME");
    }

    @Override
    public @NotNull String getID() {
        return ID;
    }

    @Override
    public @NotNull Icon getIcon() {
        return PestIcons.LOGO;
    }

    @Override
    public @Nullable PhpTestFrameworkFormDecorator getDecorator() {
        return new PhpDownloadableTestFormDecorator("https://github.com/pestphp/pest/releases");
    }
}

File: src/main/kotlin/com/pestphp/pest/PestFrameworkType.kt
package com.pestphp.pest

import com.intellij.openapi.project.Project
import com.jetbrains.php.testFramework.PhpTestFrameworkFormDecorator
import com.jetbrains.php.testFramework.PhpTestFrameworkFormDecorator.PhpDownloadableTestFormDecorator
import com.jetbrains.php.testFramework.PhpTestFrameworkType
import com.jetbrains.php.testFramework.ui.PhpTestFrameworkBaseConfigurableForm
import com.jetbrains.php.testFramework.ui.PhpTestFrameworkBySdkConfigurableForm
import com.jetbrains.php.testFramework.ui.PhpTestFrameworkConfigurableForm
import com.pestphp.pest.configuration.PestVersionDetector
import org.jetbrains.annotations.Nls
import org.jetbrains.annotations.NonNls
import javax.swing.Icon

/**
 * Registers a framework type for PHP.
 *
 * This class is used to show the menu in
 * `Preferences -> Languages & Frameworks -> PHP -> Test Frameworks`
 */
class PestFrameworkType : PhpTestFrameworkType() {
    private val pestUrl = "https://github.com/pestphp/pest/releases"

    override fun getDisplayName(): @Nls String {
        return PestBundle.message("FRAMEWORK_NAME")
    }

    override fun getID(): String {
        return ID
    }

    override fun getIcon(): Icon {
        return PestIcons.Logo
    }

    override fun getDecorator(): PhpTestFrameworkFormDecorator {
        return object : PhpDownloadableTestFormDecorator(pestUrl) {
            override fun decorate(
                project: Project,
                form: PhpTestFrameworkBaseConfigurableForm<*>
            ): PhpTestFrameworkConfigurableForm<*> {
                if (form !is PhpTestFrameworkBySdkConfigurableForm) {
                    form.setVersionDetector(PestVersionDetector.instance)
                }
                return super.decorate(project, form)
            }
        }
    }

    companion object {
        @NonNls
        val ID = "Pest"
        val instance: PhpTestFrameworkType
            get() = getTestFrameworkType(ID)
    }
}
